nixos/modules/services/security/oauth2_proxy.nix

```diff
@@ -0,0 +1,533 @@
+    ${optionalString (!isNull cfg.profileURL) "--profile-url=${cfg.profileURL}"} \
+    ${optionalString (!isNull cfg.redeemURL) "--redeem-url=${cfg.redeemURL}"} \
+    ${optionalString (!isNull cfg.redirectURL) "--redirect-url=${cfg.redirectURL}"} \
+    --request-logging=${fromBool cfg.requestLogging} \
+    ${optionalString (!isNull cfg.scope) "--scope=${cfg.scope}"} \
+    ${repeatedArgs (x: "--skip-auth-regex=${x}") cfg.skipAuthRegexes} \
+    ${optionalString (!isNull cfg.signatureKey) "--signature-key=${cfg.signatureKey}"} \
+    --upstream=${cfg.upstream} \
+    ${optionalString (!isNull cfg.validateURL) "--validate-url=${cfg.validateURL}"} \
+    ${optionalString cfg.tls.enable "--tls-cert=${cfg.tls.certificate} --tls-key=${cfg.tls.key} --https-address=${cfg.tls.httpsAddress}"}
+  '';
+in
+{
+  options.services.oauth2_proxy = {
+    enable = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Whether to run oauth2_proxy.
+      '';
+    };
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.oauth2_proxy;
+      description = ''
+        The package that provides oauth2_proxy.
+      '';
+    };
+
+    ##############################################
+    # PROVIDER configuration
+    provider = mkOption {
+      type = types.enum [
+        "google"
+        "github"
+        "azure"
+        "gitlab"
+        "linkedin"
+        "myusa"
+      ];
+      default = "google";
+      description = ''
+        OAuth provider
+      '';
+    };
+
+    approvalPrompt = mkOption {
+      type = types.enum ["force" "auto"];
+      default = "force";
+      description = ''
+        OAuth approval_prompt
+      '';
+    };
+
+    clientID = mkOption {
+      type = types.str;
+      description = ''
+        MANDATORY. The OAuth Client ID.
+      '';
+      example = "123456.apps.googleusercontent.com";
+    };
+
+    clientSecret = mkOption {
+      type = types.str;
+      description = ''
+        The OAuth Client Secret.
+      '';
+    };
+
+    skipAuthRegexes = mkOption {
+     type = types.listOf types.str;
+     default = [];
+     description = ''
+       List of regular expressions which will bypass authentication when
+       requests path's match
+     '';
+    };
+
+    # XXX: Not clear whether these two options are mutually exclusive or not.
+    email = {
+      domains = mkOption {
+        type = types.listOf types.str;
+        default = [];
+        description = ''
+          Authenticate emails with the specified domains. Use * to authenticate any email.
+        '';
+      };
+
+      addresses = mkOption {
+        type = types.nullOr types.lines;
+        default = null;
+        description = ''
+          Line-separated email addresses that are allowed to authenticate.
+        '';
+      };
+    };
+
+    loginURL = mkOption {
+      type = types.nullOr types.str;
+      default = null;
+      description = ''
+        Authentication endpoint.
+
```

Comment:
Noted. I think the newlines here make it easier to read in the file. Any objection to me leaving them as is?